lib/Sema/CSRanking.cpp

```diff
@@ -1381,3 +1353,61 @@ SolutionDiff::SolutionDiff(ArrayRef<Solution> solutions) {
     }
   }
 }
+
+InputMatcher::InputMatcher(const ValueDecl *funcDecl,
+    const ArrayRef<AnyFunctionType::Param> params) :
+    NumSkippedParameters(0), Params(params) {
+  DefaultValueMap = computeDefaultMap(params, funcDecl,
+      funcDecl->getDeclContext()->isTypeContext());
+}
+
+InputMatcher::Result InputMatcher::solve(int numInputs,
+    bool ignoreLastPair,
+    std::function<bool(unsigned, unsigned)> pairMatcher) {
```

Comment:
It would be great if instead of indices callback received references to `AnyFunctionType::Parameter`, because index itself is inconsequential here.

Reply:
I was thinking the same, Pavel. However, I found it not easy to [check for IUO](https://github.com/apple/swift/pull/17093/files#diff-a72024f35020a11c7f1e09bcd35b22bdR649) without indices. I think indices is not a bad idea, since it allows `pairMatcher` to take more control. What do you think?